Sources/NIOPosix/Bootstrap.swift

```diff
@@ -2108,61 +2124,150 @@ extension NIOPipeBootstrap {
                      "must be distinct and both positive integers.")
+        precondition(!(input == nil && output == nil), "Either input or output has to be set")
         let eventLoop = group.next()
         let channelOptions = self._channelOptions
-        try self.validateFileDescriptorIsNotAFile(input)
-        try self.validateFileDescriptorIsNotAFile(output)
 
-        let channelInitializer = { (channel: Channel) -> EventLoopFuture<ChannelInitializerResult> in
-            let initializer = self.channelInitializer ?? { _ in eventLoop.makeSucceededFuture(()) }
-            return initializer(channel).flatMap { channelInitializer(channel) }
+        let channel: PipeChannel
+        let inputFileHandle: NIOFileHandle?
+        let outputFileHandle: NIOFileHandle?
+        do {
+            if let input = input {
+                try self.validateFileDescriptorIsNotAFile(input)
+            }
+            if let output = output {
+                try self.validateFileDescriptorIsNotAFile(output)
+            }
+
+            inputFileHandle = input.flatMap { NIOFileHandle(descriptor: $0) }
+            outputFileHandle = output.flatMap { NIOFileHandle(descriptor: $0) }
+            channel = try PipeChannel(
+                eventLoop: eventLoop as! SelectableEventLoop,
+                inputPipe: inputFileHandle,
+                outputPipe: outputFileHandle
+            )
+        } catch {
+            return eventLoop.makeFailedFuture(error)
         }
 
-        let inputFileHandle = NIOFileHandle(descriptor: input)
-        let outputFileHandle = NIOFileHandle(descriptor: output)
-        let channel = try PipeChannel(
-            eventLoop: eventLoop as! SelectableEventLoop,
-            inputPipe: inputFileHandle,
-            outputPipe: outputFileHandle
-        )
 
         @Sendable
-        func setupChannel() -> EventLoopFuture<PostRegistrationTransformationResult> {
+        func setupChannel() -> EventLoopFuture<ChannelInitializerResult> {
             eventLoop.assertInEventLoop()
             return channelOptions.applyAllChannelOptions(to: channel).flatMap { _ -> EventLoopFuture<ChannelInitializerResult> in
                 channelInitializer(channel)
             }.flatMap { result in
                 eventLoop.assertInEventLoop()
                 let promise = eventLoop.makePromise(of: Void.self)
                 channel.registerAlreadyConfigured0(promise: promise)
-                return promise.futureResult.flatMap { postRegisterTransformation(eventLoop, result) }
+                return promise.futureResult.map { result }
+            }.flatMap { result -> EventLoopFuture<ChannelInitializerResult> in
+                if inputFileHandle == nil {
+                    return channel.close(mode: .input).map { result }
+                } else {
+                    return channel.selectableEventLoop.makeSucceededFuture(result)
+                }
+            }.flatMap { result in
```

Comment:
Can we save a `flatMap` here? We know either there must be at least one input file handle so we can simplify this a little.